lib/MusicBrainz/Server.pm

```diff
@@ -480,6 +506,67 @@ has json_utf8 => (
     }
 );
 
+sub form_posted_and_valid {
+    my ($self, $form, $params) = @_;
+
+    return 0 unless $self->form_posted;
+    return $self->form_submitted_and_valid($form, $params);
+}
+
+sub form_submitted_and_valid {
+    my ($self, $form, $params) = @_;
+
+    $params = $self->req->params
+        unless defined $params;
+
+    return 0 unless
+        %{$params} &&
+        $form->process(params => $params) &&
+        $form->has_params;
+
+    return 0 if
+        exists $self->action->attributes->{CSRFToken} &&
+        !$self->validate_csrf_token;
```

Comment:
`$c->validate_csrf_token` is now called as part of `$c->form_submitted_and_valid`. Hence why a lot of `$form->submitted_and_valid` checks have changed to that.

Note: The form still has to be processed first, otherwise submitted fields won't be preserved.

Reply:
Ok, `$self->action->attributes->{CSRFToken}` is checked to validate forms that don’t need CSRF token yet since it is defined server-side. :+1: